### PR TITLE
Feature/issue 333 CoR

### PR DESCRIPTION
**Pull request by maganti07**

### CODE REVIEW

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformApplyS3Handler.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      try {
+        String planFile = properties.getProperty("plan");
+        String bucketName = planFile.split("/")[2];
+        String fileName = planFile.replaceAll("s3://" + bucketName + "/", "");
+        executable.execute(String.format("aws s3api get-object --bucket %1$s --key %2$s %3$s", bucketName, fileName, planFile.substring(planFile.lastIndexOf("/")).replaceAll("/", "")));
+      } catch (InterruptedException | IOException e) {
+        e.printStackTrace();
+      }
+    } else {
+      handleRequest(properties);
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformApplyS3Handler.java]: handleRequest shouldn't be in else; the chain will break if the condition is true. Is that really what you want? Only one request to be serviced by the chain?

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  private TerraformHandler nextHandlerAction;
+
+  public TerraformHandler nextHandlerAction(TerraformHandler nextHandlerAction) {
+    this.nextHandlerAction = nextHandlerAction;
+    return nextHandlerAction;
+  }
+
+  void handleRequest(Properties properties) {
+    if (nextHandlerAction != null) {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java]: Consider using an Optional here. Null equality tests are not a great practice, in general.

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+
+/**
+ * Chain of responsibility handler.
+ */
+public abstract class TerraformHandler {
+
+  private TerraformHandler nextHandlerAction;
+
+  public TerraformHandler nextHandlerAction(TerraformHandler nextHandlerAction) {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java]: as a general rule, methods should be verbs; they are actions

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    this.nextHandlerAction = nextHandlerAction;
+    return nextHandlerAction;
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java]: This is a Handler, not an action. doAction is the method implemented, but TerraformHandler is the type. The name is confusing - by the name one would assume that its of type HandlerAction. nextHandler would be a better name.

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  void handleRequest(Properties properties) {
+    if (nextHandlerAction != null) {
+      nextHandlerAction.doAction(properties);
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandler.java]: The logic here is a little confusing. To initiate the chain, you call handleRequest, which in turn calls doAction on the next handler action (what about the current handler action?) and then the doAction implemented by the subclass is responsible for calling back to handleRequest?

Wouldn't it be simpler to abstract the execution of the action in the handler - and that's it. Then, the TerraformHandler handleRequest method is responsible for validating the next link in the chain, executing the action and moving to the next link in the chain?

Example:

```java
protected final void handleRequest(Properties properties) {
   doAction(properties);
   Optional.ofNullable(nextHandler).ifPresent(handler -> handler.handleRequest(properties));
}
```

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformPlanS3Handler.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          executable.execute(String.format("aws s3 cp %1$s %2$s --sse aws:kms --sse-kms-key-id %3$s", fileName, planFile, properties.getProperty("kmsKeyId")));
+        } else {
+          LOGGER.debug("uploading plan files to Amazon S3 with default encryption (SSE:256)");
+          executable.execute(String.format("aws s3 cp %1$s %2$s", fileName, planFile));
+        }
+      } catch (InterruptedException | IOException e) {
+        e.printStackTrace();
+      }
+    } else {
+      handleRequest(properties);
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformPlanS3Handler.java]: See the above comments. You are repeating yourself here. Think DRYer :)

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Properties;
+
+
+public class TerraformChainHandler {
+
+  private Logger logger;
+
+  String tfRootDir;
+
+  private Executable executable;
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java]: I don't think you need this 'executable'

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import com.deliveredtechnologies.io.Executable;
+import com.deliveredtechnologies.terraform.TerraformException;
+import com.deliveredtechnologies.terraform.TerraformUtils;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.Properties;
+
+
+public class TerraformChainHandler {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java]: 1. Javadocs - don't forget them ;)
2. Change the name of TerraformChainHandler to something the represents what it's used for
3. Dynamically create the chain (reflection); consider their may be pre-processing handlers and post-processing handlers

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  public TerraformChainHandler(Executable executable, Logger logger) {
+    this.executable = executable;
+    this.logger = logger;
+  }
+
+  /**
+   * Executes given planAction for a given backendType - GET/PUTs terraform plan files from/to specified backend.
+   * @param properties property options
+   */
+  public void chainInitiator(Properties properties) throws IOException, TerraformException {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformChainHandler.java]: minor: methods should generally be verbs

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformApplyS3Handler.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package com.deliveredtechnologies.terraform.terraformhandler;
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformApplyS3Handler.java]: minor: redundant package name terraform/terraformhandler
consider terraform/handler

Something that you could do for pre-processing handlers is put them in a specific package, like terraform/handler/pre

**Comment by Clayton7510** [comment]
Checkstyle errors are failing the build :'(

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformChainHandler.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  private Set<Class<? extends TerraformHandler>> getHandlerClasses() {
+    Reflections reflections = new Reflections(this.getClass().getPackage());
+    return reflections.getSubTypesOf(TerraformHandler.class);
+  }
+
+  /**
+   * Executes given planAction for a given backendType - GET/PUTs terraform plan files from/to specified backend.
+   * @param properties property options
+   */
+  public void chainInitiator(Properties properties) throws TerraformHandlerException {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformChainHandler.java]: Methods should be verbs. So, maybe runChain() or initiateChain()

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformApplyS3Handler.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.deliveredtechnologies.io.CommandLine;
+import com.deliveredtechnologies.io.Executable;
+import com.deliveredtechnologies.terraform.TerraformException;
+import com.deliveredtechnologies.terraform.TerraformUtils;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.Properties;
+
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformApplyS3Handler.java:11]: Javadocs? What's going on with the linting?

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformApplyS3Handler.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  @Override
+  public void doAction(Properties properties) {
+    if (properties.containsKey("plan") && properties.getProperty("plan").startsWith("s3")) {
+      try {
+        String planFile = properties.getProperty("plan");
+        String bucketName = planFile.split("/")[2];
+        String fileName = planFile.replaceAll("s3://" + bucketName + "/", "");
+        executable.execute(String.format("aws s3api get-object --bucket %1$s --key %2$s %3$s", bucketName, fileName, planFile.substring(planFile.lastIndexOf("/")).replaceAll("/", "")));
+      } catch (InterruptedException | IOException e) {
+        e.printStackTrace();
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformApplyS3Handler.java]: This should be handled better - let's supply a better error message.

Maybe something like ```logger.error("Unable to get " + s3KeyName + "; " + e.getMessage());```

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformPlanS3Handler.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        String fileName = planFile.substring(planFile.lastIndexOf("/")).replaceAll("/", "");
+        if (properties.containsKey("kmsKeyId")) {
+          LOGGER.debug("uploading plan files to Amazon S3 with kms encryption");
+          executable.execute(String.format("aws s3 cp %1$s %2$s --sse aws:kms --sse-kms-key-id %3$s", fileName, planFile, properties.getProperty("kmsKeyId")));
+        } else {
+          LOGGER.debug("uploading plan files to Amazon S3 with default encryption (SSE:256)");
+          executable.execute(String.format("aws s3 cp %1$s %2$s", fileName, planFile));
+        }
+      } catch (InterruptedException | IOException e) {
+        e.printStackTrace();
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformPlanS3Handler.java]: Do not handle exceptions with printStackTrace(). Instead, if you want to handle the error by logging it, use the logger.

File: tf-build-tools/tf-cmd-api/src/test/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+   */
+  @Before
+  public void setup() {
+    properties = new Properties();
+    executable = Mockito.mock(Executable.class);
+    logger = Mockito.mock(Logger.class);
+  }
+
+  @Test
+  public void executeHandlerWithDoAction() {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/test/java/com/deliveredtechnologies/terraform/terraformhandler/TerraformHandlerTest.java:32]: Two scenarios: 
1. nextHandler is present 
   set the next Handler
   call handleRequest
   validate that doAction was called on the current Handler
   validate that doAction was called on the next Handler
2. nextHandler is not present.
   set the next Handler
   call handleRequest
   validate that doAction was called on the current Handler

**Review by Clayton7510 — CHANGES_REQUESTED** [review]
Some minor updates. But it looks pretty good.

**Comment by Clayton7510** [comment]
Please remove the tfplan files from the PR.

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformChainHandler.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Properties;
+import java.util.Set;
+
+public class TerraformChainHandler {
+
+  private Logger logger;
+
+  String tfRootDir;
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformChainHandler.java:20]: Why not are the instance variables package private?

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformHandlerException.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package com.deliveredtechnologies.terraform.handler;
+
+public class TerraformHandlerException extends Exception {
```
Comment (Clayton7510) [inline tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/terraform/handler/TerraformHandlerException.java:6]: JavaDocs, pls

**Review by Clayton7510 — APPROVED** [review]

**Comment by palaniswamy** [comment]
All good

**Review by palaniswamy — APPROVED** [review]
Approved

**Pull request merged**